Add gradients and valueAndGradients. Remove gradientWrt on the public math API.

nsthorat opened the pull request:
This PR:
* Adds top-level `gradients` and `valueAndGradients` method to NDArrayMath. These both take a function that returns a Scalar value, and a second parameter of what to compute gradients with respect to. These start a gradients scope, execute the function, and compute the derivative of y with respect to given inputs.
* Adds top-level `gradientsScope` which forces all inner scopes not to clean anything up.
* Flattens all tape structures (no more subtapes). There is only 1 tape when a gradientScope begins, and all kernels get enqueued on the tape. It gets thrown out when the last gradient scope ends.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/462)
<!-- Reviewable:end -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/math/element_wise_arithmetic_test.ts, line 234 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L19whT1-SdeNvFkWgkM:-L19whT1-SdeNvFkWgkN:bem1o29) ([raw file](https://github.com/pair-code/deeplearnjs/blob/62f60f44b6d195f8931ce2fbb55954aac53c568f/src/math/element_wise_arithmetic_test.ts#L234)):*
> ```XML
> {
>   const tests: MathTests = it => {
>     it('Broadcast throws', math => {
> ```

just double checking that the idea is to re-add this test later when grad works with broadcast, right? any particular reason why removing it now?

---

*[src/math/element_wise_arithmetic_test.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DV1lNhmZcFPlpvZ1N:-L1DV1lNhmZcFPlpvZ1O:bhtol4x) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/element_wise_arithmetic_test.ts#L238)):*
> valueAndGradients

nit: cleaner to unpack directly, here and elsewhere in this PR. Code will look much cleaner and reduce line wrapping.

`const {value, gradients} = math.valueAndGradients(() => ...`

---

*[src/math/element_wise_arithmetic_test.ts, line 276 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DWEOCghSAvtr7BSGq:-L1DWEOCghSAvtr7BSGr:b-kxhqs) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/element_wise_arithmetic_test.ts#L276)):*
> 1e-1

can you remind me again why the byte packing texture gets such bad accuracy? I'm assuming this threshold is due to byte packing imprecision. 

---

*[src/math/math.ts, line 960 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DXIbVay4euY1I9ngE:-L1DXIbVay4euY1I9ngF:b93d9dc) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math.ts#L960)):*
> ```XML
>     const gradient = (dy: NDArray<G>, y: NDArray<G>) => {
>       if (!util.arraysEqual(a.shape, b.shape)) {
>         throw new Error(
> ```

let's file an issue to track gradients for broadcasted shapes (you can enumerate all the ops that need it)

---

*[src/math/math.ts, line 2344 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DYoM12GLxGmhIg9O-:-L1DYoM12GLxGmhIg9O0:bz44zgt) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math.ts#L2344)):*
> flattenNameArrayMap

missed this in the previous PR: Object.keys() by JS specs can be non-deterministic - browser implementations probably are deterministic, but just to be safe, make flattenNameArrayMap() return both the keys, and the flattened array, and pass those keys later to unflatten

---

*[src/math/math_test.ts, line 370 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DZkGhntGY1pMu2G5I:-L1DZkGhntGY1pMu2G5J:b-5b8gpa) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math_test.ts#L370)):*
> is

typo: `s/is/if/`

---

*[src/math/math_test.ts, line 429 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1D_0K9wAM_XmEFRoSG:-L1D_0K9wAM_XmEFRoSH:bia4k89) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math_test.ts#L429)):*
> 2

nit: `s/2/a.get()` makes it a little more explicit that the answer is 6a. Here and the test below

---

*[src/math/math_test.ts, line 445 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1D_O-Y5akO84C4bMTm:-L1D_O-Y5akO84C4bMTn:b-kh78ox) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math_test.ts#L445)):*
> ```XML
>       expect(gradients.shape).toEqual(a.shape);
>       test_util.expectArraysClose(
>           gradients.dataSync(), new Float32Array([6 * 2]));
> ```

add a todo(smilkov|nsthorat) to add expect(math.getNumArrays()) before and after once I fix math.getNumArrays(). Would be great to test for memory cleanup in all these integration tests. Good to file an issue as well

---

*[src/math/matmul_test.ts, line 325 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1Da03rbSlv0hvfCnVR:-L1Da03rbSlv0hvfCnVS:b6479bg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/matmul_test.ts#L325)):*
> valueAndGradients

since you call valueAndGradients, might as well test for the value itself, not just the gradients.

---

*[src/math/unaryop_test.ts, line 122 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DbcjAE-YMWYjloXzf:-L1DbcjAE-YMWYjloXzg:bjd6xw2) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/unaryop_test.ts#L122)):*
> expectArraysClose

I know it's a bit late in the game :) , but reading all these tests will be SO much easier and concise if we allow expectArraysClose to take NDArray|TypedArray|number[], and test for the values internally.

If one of the arguments is number[], we won't test for dtype matching - in that case we can ask the test authors to tests for dtype explicitly.

We can keep the old behavior for backwards compatibility.

WDYT? Worth filing an issue?

---

*[src/math/backends/backend_engine.ts, line 80 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DgBFDROYZYMIz8HeI:-L1DgBFDROYZYMIz8HeJ:b-pzbu1a) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/backends/backend_engine.ts#L80)):*
> ```XML
>     }
> 
>     config = tape_util.stripUndefinedInputsFromInputConfig(config) as C;
> ```

move the config line inside `if (this.activeTape != null)`

---

*[src/math/backends/backend_engine.ts, line 166 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DhOmfzsY_XUY8Twgx:-L1DhOmfzsY_XUY8Twgy:b-x3gozu) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/backends/backend_engine.ts#L166)):*
> this.track(ndarray)

not sure when this changed, but can you make it (again) a no-op. i.e.

`const trackFn = <T extends NDArray>(ndarray: T): T => {}`

---

*[src/math/backends/tape_util.ts, line 194 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DdHToexVwSi6xHYMd:-L1DdHToexVwSi6xHYMe:b-lnkx3y) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/backends/tape_util.ts#L194)):*
> // Call the backend directly so we don't add the "add" node to our
>         // tape.

This comment is out of sync with the code change. I'm assuming you wanted `add` to be part of the tape so 2nd order gradient works, correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-:-L1Drj5EuE1XXQxulyf9:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Thanks for the thorough review!

---

Review status: 1 of 12 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[src/math/element_wise_arithmetic_test.ts, line 234 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L19whT1-SdeNvFkWgkM:-L1FeXocHVdfuNnlJW67:bqk4w49) ([raw file](https://github.com/pair-code/deeplearnjs/blob/62f60f44b6d195f8931ce2fbb55954aac53c568f/src/math/element_wise_arithmetic_test.ts#L234)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

just double checking that the idea is to re-add this test later when grad works with broadcast, right? any particular reason why removing it now?
</blockquote></details>

Removing it because I think it's kind of a useless test... going to add back proper broadcasting tests once we add it.

---

*[src/math/element_wise_arithmetic_test.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DV1lNhmZcFPlpvZ1N:-L1FfjIRTJHevCIucaOS:b-l2brxz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/element_wise_arithmetic_test.ts#L238)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> valueAndGradients

nit: cleaner to unpack directly, here and elsewhere in this PR. Code will look much cleaner and reduce line wrapping.

`const {value, gradients} = math.valueAndGradients(() => ...`
</blockquote></details>

Good point, looks much better. Done.

---

*[src/math/element_wise_arithmetic_test.ts, line 276 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DWEOCghSAvtr7BSGq:-L1Ffo9o3R7gX0HYbhar:b-5la0gc) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/element_wise_arithmetic_test.ts#L276)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> 1e-1

can you remind me again why the byte packing texture gets such bad accuracy? I'm assuming this threshold is due to byte packing imprecision. 
</blockquote></details>

Yeah :(

I think here it gets poor accuracy since we are accumulating the error over several multiply adds.

---

*[src/math/math.ts, line 960 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DXIbVay4euY1I9ngE:-L1Fgvt1E4oCJJiykyVH:b-cikylk) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math.ts#L960)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

let's file an issue to track gradients for broadcasted shapes (you can enumerate all the ops that need it)
</blockquote></details>

I'm going to actually file a big issue once we're comfortable with the gradients API. I don't want to do it yet because I want to first make sure this API makes sense once it's complete and we can use it, otherwise we might have to refactor much more.

Once I file that issue we can get OSS community to contribute gradients with the new API - I will do a pass through of all kernels we have!

---

*[src/math/math.ts, line 2344 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DYoM12GLxGmhIg9O-:-L1FnnOy7Vg71M9GZWHg:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math.ts#L2344)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> flattenNameArrayMap

missed this in the previous PR: Object.keys() by JS specs can be non-deterministic - browser implementations probably are deterministic, but just to be safe, make flattenNameArrayMap() return both the keys, and the flattened array, and pass those keys later to unflatten
</blockquote></details>

Done.

---

*[src/math/math_test.ts, line 370 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DZkGhntGY1pMu2G5I:-L1Fj28qrEkWc0Hb8qh2:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math_test.ts#L370)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> is

typo: `s/is/if/`
</blockquote></details>

Done.

---

*[src/math/math_test.ts, line 429 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1D_0K9wAM_XmEFRoSG:-L1FjJbhndhutVamgXGw:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math_test.ts#L429)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> 2

nit: `s/2/a.get()` makes it a little more explicit that the answer is 6a. Here and the test below
</blockquote></details>

Done.

---

*[src/math/math_test.ts, line 445 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1D_O-Y5akO84C4bMTm:-L1FkHG8CIhz5VrKh_el:b-c54xmo) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/math_test.ts#L445)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

add a todo(smilkov|nsthorat) to add expect(math.getNumArrays()) before and after once I fix math.getNumArrays(). Would be great to test for memory cleanup in all these integration tests. Good to file an issue as well
</blockquote></details>

Done.

https://github.com/PAIR-code/deeplearnjs/issues/467

---

*[src/math/matmul_test.ts, line 325 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1Da03rbSlv0hvfCnVR:-L1FkjXnOZzY4Z-S-LkO:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/matmul_test.ts#L325)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> valueAndGradients

since you call valueAndGradients, might as well test for the value itself, not just the gradients.
</blockquote></details>

Done.

---

*[src/math/unaryop_test.ts, line 122 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DbcjAE-YMWYjloXzf:-L1G0fR6ozPBDLhIPOFD:bjzv248) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/unaryop_test.ts#L122)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> expectArraysClose

I know it's a bit late in the game :) , but reading all these tests will be SO much easier and concise if we allow expectArraysClose to take NDArray|TypedArray|number[], and test for the values internally.

If one of the arguments is number[], we won't test for dtype matching - in that case we can ask the test authors to tests for dtype explicitly.

We can keep the old behavior for backwards compatibility.

WDYT? Worth filing an issue?
</blockquote></details>

Added support for that and updated all the tests in this PR.

Filed: https://github.com/PAIR-code/deeplearnjs/issues/468

---

*[src/math/backends/backend_engine.ts, line 80 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DgBFDROYZYMIz8HeI:-L1FljNwZ1ZpQ5RRELKF:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/backends/backend_engine.ts#L80)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

move the config line inside `if (this.activeTape != null)`
</blockquote></details>

Done.

---

*[src/math/backends/backend_engine.ts, line 166 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DhOmfzsY_XUY8Twgx:-L1Flpv1yIH3w0LgUgn6:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/backends/backend_engine.ts#L166)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> this.track(ndarray)

not sure when this changed, but can you make it (again) a no-op. i.e.

`const trackFn = <T extends NDArray>(ndarray: T): T => {}`
</blockquote></details>

Done.

---

*[src/math/backends/tape_util.ts, line 194 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-L1DdHToexVwSi6xHYMd:-L1FluPAh01u0RvLY6Dw:b97akdv) ([raw file](https://github.com/pair-code/deeplearnjs/blob/eb18f1769695b4aca64717579efa6a17eade37cd/src/math/backends/tape_util.ts#L194)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> // Call the backend directly so we don't add the "add" node to our
>         // tape.

This comment is out of sync with the code change. I'm assuming you wanted `add` to be part of the tape so 2nd order gradient works, correct?

</blockquote></details>

Correct, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/462#-:-L1G150ITL9o2v255w7f:bst74u7)*
<!-- Sent from Reviewable.io -->
